### PR TITLE
Withdraw endpoint is added

### PR DESCRIPTION
**Pull request by philter87**
Withdraw Certificate endpoint is added

Changes:
- WithdrawnCertificate model is created with corresponding Repository, migration script and test
- Endpoint is added with WithdrawCertificateController and corresponding tests

### CODE REVIEW

File: src/ProjectOrigin.Stamp.Server/Services/REST/v1/WithdrawCertificateController.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        request.Transactions.Add(transaction);
+
+        using var channel = GrpcChannel.ForAddress(registryOptions.Value.GetRegistryUrl(registry));
+        var client = new RegistryService.RegistryServiceClient(channel);
+        await client.SendTransactionsAsync(request);
+
+        await unitOfWork.WithdrawnCertificateRepository.Create(registry, certificateId);
+        unitOfWork.Commit();
+
+        return Accepted(new WithdrawnCertificateResponse
```
Comment (tnickelsen) [inline src/ProjectOrigin.Stamp.Server/Services/REST/v1/WithdrawCertificateController.cs:64]: Lav denne til Created. Accepted er til 'pending' requests

File: src/ProjectOrigin.Stamp.Test/Repositories/WithdrawnCertificateRepositoryTests.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        connection.Open();
+        _repository = new WithdrawnCertificateRepository(connection);
+    }
+
+    [Fact]
+    public async Task CreateAndGetCertificate()
+    {
+        // Arrange
+        var certificateId = Guid.NewGuid();
+        var registryName = "Energinet.dk";
```
Comment (tnickelsen) [inline src/ProjectOrigin.Stamp.Test/Repositories/WithdrawnCertificateRepositoryTests.cs:25]: Use fictional names (Narnia fx) for this or random strings. We are open source and should not have energinet specific code

File: src/ProjectOrigin.Stamp.Test/CertificateIssuingTests.cs [hunk 3]
```diff
@@ -24,171 +28,70 @@ public CertificateIssuingTests(TestServerFixture<Startup> fixture, PostgresDatab
         fixture.PostgresConnectionString = postgres.ConnectionString;
         fixture.RabbitMqOptions = rabbitMq.Options;
         fixture.RegistryOptions = poStack.RegistryOptions;
+
+        _gridArea = fixture.RegistryOptions.IssuerPrivateKeyPems.First().Key;
+        _registryName = fixture.RegistryOptions.Registries[0].Name;
+        _walletClient = _poStack.CreateWalletClient(Guid.NewGuid().ToString());
```
Comment (tnickelsen) [inline src/ProjectOrigin.Stamp.Test/CertificateIssuingTests.cs:34]: Nice cleanup af disse tests. Måske lige lav et dispose af _walletClient efter hver test

File: src/ProjectOrigin.Stamp.Test/CertificateIssuingTests.cs [hunk 4]
```diff
@@ -252,62 +133,97 @@ public async Task IssueCertificate(CertificateType type)
...
-                { "assetId", gsrn },
+                { "assetId", _gsrn },
                 { "fuelCode", "F01040100" },
                 { "techCode", "T010000" }
             });
         }
     }
+
+    [Fact]
+    public async Task WithdrawCertificateSuccess()
```
Comment (tnickelsen) [inline src/ProjectOrigin.Stamp.Test/CertificateIssuingTests.cs:176]: Denne ville jeg bare kalde 'WithdrawCertificate()'

Comment (tnickelsen) [reply src/ProjectOrigin.Stamp.Test/CertificateIssuingTests.cs:176]: Læg disse WithdrawCertificate tests i en klasse for sig og brug samme fixtures, så de kun spindes op én gang. Du kan se et eksempel på, hvordan man gør det her: https://github.com/Energinet-DataHub/energy-origin/blob/main/domains/certificates/Query.API/API.IntegrationTests/IntegrationTestCollection.cs

Og her for brug heraf: https://github.com/Energinet-DataHub/energy-origin/blob/main/domains/certificates/Query.API/API.IntegrationTests/CertificateIssuingTests.cs

File: src/ProjectOrigin.Stamp.Test/CertificateIssuingTests.cs [hunk 5]
```diff
@@ -252,62 +133,97 @@ public async Task IssueCertificate(CertificateType type)
...
+        // Act
+        var responseWithdraw = await _client.WithdrawCertificate(_registryName, cert.Id);
+
+        // Assert
+        responseWithdraw.StatusCode.Should().Be(HttpStatusCode.Accepted);
+    }
+
+
+    [Fact]
+    public async Task WithdrawCertificateShouldFailIfNoCertificateExistsInDatabase()
```
Comment (tnickelsen) [inline src/ProjectOrigin.Stamp.Test/CertificateIssuingTests.cs:192]: Kald 'WhenNoCertificateExistsInDatabase_NotFound()'

File: src/ProjectOrigin.Stamp.Test/CertificateIssuingTests.cs [hunk 6]
```diff
@@ -252,62 +133,97 @@ public async Task IssueCertificate(CertificateType type)
...
+        // Act
+        var responseWithdraw = await _client.WithdrawCertificate(_registryName, cert.Id);
+
+        // Assert
+        responseWithdraw.StatusCode.Should().Be(HttpStatusCode.NotFound);
+    }
+
+
+    [Fact]
+    public async Task WithdrawCertificateShouldFailIfAlreadyWithdrawn()
```
Comment (tnickelsen) [inline src/ProjectOrigin.Stamp.Test/CertificateIssuingTests.cs:206]: Kald 'WhenCertificateAlreadyWithdrawn_Conflict()'

File: src/ProjectOrigin.Stamp.Server/Helpers/Registry.cs [hunk 7]
```diff
@@ -85,6 +85,32 @@ public static Transaction CreateTransaction(this IssuedEvent issuedEvent, IPriva
...
+    public static Transaction CreateTransaction(this WithdrawnEvent withdrawnEvent, string registryName, Guid certificateId, IPrivateKey issuerKey)
+    {
+        var header = new TransactionHeader
+        {
+            FederatedStreamId = new Common.V1.FederatedStreamId
+            {
+                Registry = registryName,
+                StreamId = new Common.V1.Uuid { Value = certificateId.ToString() }
+            },
+            PayloadType = IssuedEvent.Descriptor.FullName,
```
[the remark this answers is not in the document]
Comment (philter87) [reply src/ProjectOrigin.Stamp.Server/Helpers/Registry.cs:97]: Godt set! og umiddelbart ret vigtig. Mine test burde vel egentlig fejle

[the remark this answers is not in the document]
Comment (tnickelsen) [reply src/ProjectOrigin.Stamp.Server/Helpers/Registry.cs:97]: Hmm om registriet burde kaste op, ved jeg faktisk ikke. Det må du snakke med Martin Schmidt om. Men man kunne måske godt query registriet i testen og sige, man forventer, der ligger et withdrawn event

File: src/ProjectOrigin.Stamp.Server/Services/REST/v1/WithdrawCertificateController.cs [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        request.Transactions.Add(transaction);
+
+        using var channel = GrpcChannel.ForAddress(registryOptions.Value.GetRegistryUrl(registry));
+        var client = new RegistryService.RegistryServiceClient(channel);
+        await client.SendTransactionsAsync(request);
+
+        await unitOfWork.WithdrawnCertificateRepository.Create(registry, certificateId);
+        unitOfWork.Commit();
+
+        return Created($"v1/certificates/{registry}/{certificateId}", new WithdrawnCertificateResponse
```
Comment (tnickelsen) [inline src/ProjectOrigin.Stamp.Server/Services/REST/v1/WithdrawCertificateController.cs:64]: This url does not go anywhere, so just return Created without a url.

Comment (philter87) [reply src/ProjectOrigin.Stamp.Server/Services/REST/v1/WithdrawCertificateController.cs]: Det er rigtigt :)

**Review by tnickelsen — APPROVED** [review]

**Pull request merged**